src/main/java/gregtech/api/capability/impl/MultiblockRecipeLogic.java

```diff
@@ -71,13 +85,76 @@ protected IMultipleTankHandler getOutputTank() {
             return;
+        }
+
+        // Distinct buses only apply to some of the multiblocks, so check the controller against a lower class
+        if(controller instanceof RecipeMapMultiblockController) {
+            RecipeMapMultiblockController distinctController = (RecipeMapMultiblockController) controller;
+
+            if(distinctController.canBeDistinct() && distinctController.isDistinct()) {
+                trySearchNewRecipeDistinct();
+                return;
+            }
+        }
 
         super.trySearchNewRecipe();
     }
 
+    protected void trySearchNewRecipeDistinct() {
+        long maxVoltage = getMaxVoltage();
+        Recipe currentRecipe = null;
+        List<IItemHandlerModifiable> importInventory = getInputBuses();
+        IMultipleTankHandler importFluids = getInputTank();
+
+        // Our caching implementation
+        // This guarantees that if we get a recipe cache hit, our efficiency is no different from other machines
+        if (previousRecipe != null && previousRecipe.matches(false, importInventory.get(lastRecipeIndex), importFluids)) {
+            currentRecipe = previousRecipe;
+            // TODO, is setting the invalidInputs here needed?
+            invalidInputsForRecipes = false;
+            // If a valid recipe is found, immediately attempt to return it to prevent inventory scanning
+            if (setupAndConsumeRecipeInputs(currentRecipe, importInventory.get(lastRecipeIndex))) {
+                setupRecipe(currentRecipe);
+                metaTileEntity.getNotifiedItemInputList().clear();
+                metaTileEntity.getNotifiedFluidInputList().clear();
+                //TODO, should we set the previous recipe to the current recipe here, or would that cause some issues with caching once parallel is introduced?
+                return;
+            }
+        }
+
+        // On a cache miss, our efficiency is much worse, as it will check
+        // each bus individually instead of the combined inventory all at once.
+        for (int i = 0; i < importInventory.size(); i++) {
+            IItemHandlerModifiable bus = importInventory.get(i);
+            boolean inputsChanged = hasNotifiedInputs();
+            // If the inputs have changed since the last recipe calculation, find a new recipe based on the new inputs
+            if (inputsChanged) {
+                currentRecipe = findRecipe(maxVoltage, bus, importFluids, MatchingMode.DEFAULT);
+                // Cache the current recipe, if one is found
+                if (currentRecipe != null) {
+                    this.previousRecipe = currentRecipe;
+                }
+            }
+
+            invalidInputsForRecipes = (currentRecipe == null);
+
+            if (currentRecipe != null && setupAndConsumeRecipeInputs(currentRecipe, importInventory.get(i))) {
+                lastRecipeIndex = i;
+                setupRecipe(currentRecipe);
+                metaTileEntity.getNotifiedItemInputList().clear();
```

Comment:
again, only remove the current bus from the list

Reply:
Can we be sure that the buses are ordered consistently, and that we will never get index out of bounds exceptions if we target a specific bus?